<a href="https://colab.research.google.com/github/ksh001/ksh0011/blob/main/py_04_pdS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Pandas Study 03 복습 

In [206]:
import pandas as pd

df = pd.DataFrame(
    {'a' : [3, 2, 3, 5, None],
     'b' : [45, 60, 9, 100, 12],
     'c' : [4, 203, 123, 7, 4],
     'd' : [45, 2, 9, 100, 12],
     'e' : [4, 203, 123, 7, None]},
   index = [1, 2, 3, 4, 5])

df

a    b    c    d      e
1  3.0   45    4   45    4.0
2  2.0   60  203    2  203.0
3  3.0    9  123    9  123.0
4  5.0  100    7  100    7.0
5  NaN   12    4   12    NaN

In [8]:
df.dropna() #none을 drop

a    b    c    d
1  3.0   45    4   45
2  3.0    2   23    2
3  4.0    9  123    9
4  5.0  100    3  100

In [209]:
df.fillna(3) #none에 값 대입

a    b    c    d      e
1    3.0   45    4   45    4.0
2    2.0   60  203    2  203.0
3    3.0    9  123    9  123.0
4    5.0  100    7  100    7.0
5  400.0   12    4   12    3.0

In [208]:
df.loc[5, 'a'] = 400
df

a    b    c    d      e
1    3.0   45    4   45    4.0
2    2.0   60  203    2  203.0
3    3.0    9  123    9  123.0
4    5.0  100    7  100    7.0
5  400.0   12    4   12    NaN

##qcut
    특정 행을 n등분해서 구간을 같은 개수로 구간으로 나눠준다. 구간의 이름을 설정하지 않으면(==False) 0123 등으로 구분한다.
    수로 이루어진 값을 구간을 나누어 구역으로 문자화 할 때 씀

In [137]:
pd.qcut(df.d, 5, labels=False) 

1    3
2    0
3    1
4    4
5    2
Name: d, dtype: int64

In [214]:
pd.qcut(df.d, 5) 

1     (25.2, 56.0]
2     (1.999, 7.6]
3      (7.6, 10.8]
4    (56.0, 100.0]
5     (10.8, 25.2]
Name: d, dtype: category
Categories (5, interval[float64]): [(1.999, 7.6] < (7.6, 10.8] < (10.8, 25.2] < (25.2, 56.0] <
                                    (56.0, 100.0]]

In [213]:
pd.qcut(df.a, 3, labels=['small', 'medium', 'big'])

1    small
2    small
3    small
4      big
5      big
Name: a, dtype: category
Categories (3, object): ['small' < 'medium' < 'big']

In [61]:
pd.qcut(df.b, 4, labels=['s', 'm1', 'm2', 'b'])

1    m1
2    m2
3     s
4     b
5     s
Name: b, dtype: category
Categories (4, object): ['s' < 'm1' < 'm2' < 'b']

In [138]:
pd.qcut(df.d, 10, labels=False) 

1    7
2    0
3    2
4    9
5    4
Name: d, dtype: int64

###qcut의 오류 원인

You can drop duplicate edges by setting the 'duplicates' kwarg

https://stackoverflow.com/questions/20158597/how-to-qcut-with-non-unique-bin-edges

  The problem is that pandas.qcut chooses the bins/quantiles so that each one has the same number of records, but all records with the same value must stay in the same bin/quantile (this behaviour is in accordance with the statistical definition of quantile).

    pandas.qcut은 각 기록이 동일한 수의 레코드를 가질 수 있도록 bin/누적백분율를 선택하는데, 값이 동일한 모든 레코드는 동일한 bin/누적백분율에 있어야 한다
    (누적백분율의 통계적 정의에 따라 작동하도록 설계돼있다).

In [194]:
pd.qcut(df.c, 3, labels=False)
#pd.qcut(df.a, 4, labels=False)

1    0
2    2
3    2
4    1
5    0
Name: c, dtype: int64

In [190]:
df

a    b    c    d      e
1  3.0   45    4   45    4.0
2  3.0   60  203    2  203.0
3  3.0    9  123    9  123.0
4  5.0  100    7  100    7.0
5  NaN   12    4   12    NaN

In [165]:
pd.qcut(df.c, 3)

1       (3.999, 5.0]
2    (84.333, 203.0]
3    (84.333, 203.0]
4      (5.0, 84.333]
5       (3.999, 5.0]
Name: c, dtype: category
Categories (3, interval[float64]): [(3.999, 5.0] < (5.0, 84.333] < (84.333, 203.0]]

    (3.999, 7.0]
    (7.0, 123.0]
    (123.0, 203.0] 처럼 될 것 같지만 이와 다름.

In [187]:
df.c.quantile([0, 0.333333333, 0.66666666, 1]) #백분율이 4를 두개 포함하는 실 백분율임

0.000000      4.00000
0.333333      5.00000
0.666667     84.33333
1.000000    203.00000
Name: c, dtype: float64

In [183]:
df.e.quantile([0, 0.333333333, 0.666666666, 1])

0.000000      4.0
0.333333      7.0
0.666667    123.0
1.000000    203.0
Name: e, dtype: float64

In [177]:
df.c.quantile([0, 0.25, 0.5, 0.75, 1]) #누적백분율이 4를 두개 포함하여 계산됐었다는 것을 알 수 있다. 4 4 7 123 203

0.00      4.0
0.25      4.0
0.50      7.0
0.75    123.0
1.00    203.0
Name: c, dtype: float64

    (3.999, 4.0]
    (4.0, 7.0] 
    (7.0, 123.0] 
    (123.0, 203.0] 

    quantile(누적백분율)의 통계적 정의를 따를 때, 4,4,7,123, 203에서 4와 4는 같은 위치에 있어야 하므로 4, 7, 123, 203으로 나누는 것이 맞다. 
    다른 코드에서도 직접적으로 코드가 오류가 안나면 맞는 것처럼 보이지만, 사실은 틀린 값이었다.

In [174]:
df.e.quantile([0, 0.25, 0.5, 0.75, 1]) #실제 4, 7, 123, 203의 백분율은 이렇게 나와야 한다. 

0.00      4.00
0.25      6.25
0.50     65.00
0.75    143.00
1.00    203.00
Name: e, dtype: float64

## 다른 코드에서도 직접적으로 코드가 오류가 안나면 맞는 것처럼 보이지만, 사실은 틀린 값이었다.

In [179]:
pd.qcut(df.c, 3, labels = False)

1    0
2    2
3    2
4    1
5    0
Name: c, dtype: int64

In [180]:
pd.qcut(df.e, 3, labels = False)

1    0.0
2    2.0
3    1.0
4    0.0
5    NaN
Name: e, dtype: float64

###   4칸일때만 오류가 나는 이유

    

    0  (3.999, 4.0] -> 이러한 구간이 만들어질 때 부터 디버깅 오류가 난다.
    1  (4.0, 7.0] 
    2  (7.0, 123.0] 
    3  (123.0, 203.0] 

In [220]:
pd.qcut(df.e, 10000)

1       (3.9999, 4.0009]
2       (202.976, 203.0]
3    (122.9768, 123.008]
4       (6.9997, 7.0232]
5                    NaN
Name: e, dtype: category
Categories (10000, interval[float64]): [(3.9999, 4.0009] < (4.0009, 4.0018] < (4.0018, 4.0027] <
                                        (4.0027, 4.0036] ... (202.904, 202.928] <
                                        (202.928, 202.952] < (202.952, 202.976] < (202.976, 203.0]]

In [110]:
#pd.qcut(df.a, 4) 오류코드

In [ ]:
#나중에 해보기
df = (pd.melt(df)
          .rename(columns={
          'variable' : '칼럼이름',
          'value' : '값'})
          
          .query('값' >= 2')
)
df

